Created FlashPaletteEffectWarhead

penev92 opened the pull request:
This is a small refactoring that decouples the `NukePower` trait (via its `NukeLaunch` projectile) from the `FlashPaletteEffect` trait and allows for better reusability.

This somewhat makes `FlashPaletteEffectInfo.Length` redundant, as well as the -1 tick special case tied to it.

Depends on #17283.

penev92 committed:
Fixed a bug with the NukeLaunch's Player palette

abcdefg30 commented:
Needs a rebase.

penev92 committed:
Updated a stale copyright header

penev92 commented:
Rebased, but I want to discuss getting rid of `FlashPaletteEffectInfo.Length` and/or the `if (ticks == -1)` case in `FlashPaletteEffect.Enable()` before this gets merged.

Mailaender commented:
Keep the duration on the warhead and the Lua API `void Flash(string type = nil, int ticks = -1)` then remove the duplication on `FlashPaletteEffectInfo.Length`.

Mailaender commented:
Removing `FlashPaletteEffectInfo.Length` would require a Lua API `void Flash(string type = nil, int ticks = -1)` change as well.

Mailaender reviewed (dismissed):
Sensible refactor.

pchote commented:
Needs a rebase now thanks to the hotfix release.

penev92 commented:
Rebased.

abcdefg30 commented on OpenRA.Mods.Common/UpdateRules/Rules/20191117/CreateFlashPaletteEffectWarhead.cs:
```diff
@@ -0,0 +1,61 @@
+		{
+			if (weaponsToUpdate.Any())
+				yield return "Add a FlashPaletteEffectWarhead to the following weapons:\n" +
+							 UpdateUtils.FormatMessageList(weaponsToUpdate.Select(x => "Weapon `{0}`, used by trait `{1}` on actor {2}".F(x.Item1, x.Item2, x.Item3)));
```
The indentation here looks off.

penev92 replied:
Fixed.

abcdefg30 commented on OpenRA.Mods.Common/Warheads/FlashPaletteEffectWarhead.cs:
```diff
@@ -0,0 +1,34 @@
+		public readonly string FlashType = null;
+
+		[Desc("Duration of the flashing, measured in ticks. Set to -1 to default to the `Length` of the `FlashPaletteEffect`.")]
+		public readonly int Duration = 0;
```
Why isn't the default `-1`? If we leave this at zero we should `FieldLoader.Require` it instead Imho.

penev92 replied:
Added `FieldLoader.Require`.

abcdefg30 commented on mods/ra/rules/structures.yaml:
```diff
@@ -51,7 +51,6 @@ MSLO:
 		DisplayBeacon: True
 		DisplayRadarPing: True
 		BeaconPoster: atomicon
-		FlashType: Nuke
```
I just noticed this change is in a different commit than it's C# counterpart, which means things would probably break if someone was to jump in the middle of the commits here, say during a bisect. I think we usually ignore such things, and I want to ask this as more general question: @reaperrr, @pchote do we want to establish any sort of ruling for how such things should be handled? (a. Ignore as before, since it is unlikely to have an impact and is easily worked around if it has, or b. force commits to always compile and run the game)

penev92 replied:
I don't think it's feasible to hope every single commit to be stable and working, as a good chunk of them will be parts of major refactorings these days.

abcdefg30 replied:
The logic doesn't need to work perfectly, but it would be very helpful and should be possible to have each commit compile and at least start the game.

Mailaender commented:
Needs a small rebase.

Smittytron reviewed (dismissed):
Works as advertised. We should think about adding a flash to mininuke and cratenuke (not necessarily in this PR)

penev92 commented:
Rebased.
I have also given up on bringing up the Lua API cleanup, so we can merge this whenever.

abcdefg30 commented:
> We should think about adding a flash to mininuke and cratenuke (not necessarily in this PR)

:+1: to adding it to those.

penev92 committed:
Added a FlashPaletteEffect to RA nukes and removed from D2k

penev92 commented:
Added the FlashEffectWarhead to the RA nuke truck and nuke crate.
Also removed from the D2k Death Hand missile.

teinarss commented:
looks good, needs a rebase and do we want some squash on the commits?

penev92 committed:
Created FlashPaletteEffectWarhead and removed the hardcoded flashing from NukeLaunch

Not actually hardcoded, but there was a hard dependency in NukeLaunch on FlashPaletteEffect and explicit flashing checks.

Mailaender commented:
@penev92 silently updated this.

abcdefg30 commented:
Actually, the update rule is in the wrong subfolder.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5bcdcf263c52b29fa1a317f8077393cb8689a296...7697162459abcbf2cde29cfd6a2cef7e577ad6fb)
